File: src/main/java/com/HOIIVUtils/clauzewitz/localization/Localizable.java
package com.HOIIVUtils.clauzewitz.localization;

import com.HOIIVUtils.clauzewitz.HOIIVFile;
import com.HOIIVUtils.ui.FXWindow;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.*;
import java.util.function.Supplier;
import java.util.stream.Collectors;

/**
 * Interface for Clausewitz-localizable objects. This interface is specifically related to
 * Clausewitz-engine localization, and not the general localization/i18n of this program.
 */
public interface Localizable {

	enum Property {
		NAME,
		DESCRIPTION,
	}

	/**
	 * Default method to get the number of localizable properties.
	 * @implNote this method should not usually be overwritten.
	 * @return the number of localizable properties.
	 */
	default int numLocalizableProperties() {
		return getLocalizableProperties().size();
	}

	/**
	 * Default method to get the localizable property identifiers and keys.
	 * @return a map of localizable property identifiers and keys.
	 */
	@NotNull Map<Property, String> getLocalizableProperties();

	default List<String> getLocalizationKeys() {
		return new ArrayList<>(getLocalizableProperties().values());
	}

	/**
	 * Add a localizable property using the specified key.
	 * @param property the localized property to add.
	 * @param localizationKey the property localization key to add.
	 */
	default void addLocalizableProperty(Property property, String localizationKey) {
		getLocalizableProperties().put(property, localizationKey);
	}

	/**
	 * Default method to clear the localizable properties.
	 */
	default void clearLocalizableProperties() {
		getLocalizableProperties().clear();
	}

	default List<Localization> getLocalization() {
		return LocalizationManager.getAll(getLocalizationKeys());
	}

	/**
	 * Gets the localization for the given property.
	 * @param property the localizable property to get localization for.
	 * @return the localization for the given property.
	 */
	default Localization localization(Property property) {
		String key = getLocalizableProperties().get(property);
		if (key == null) return null;
		return LocalizationManager.get(key);
	}

	/**
	 * Gets the localization text for the given property.
	 * @param property the localizable property to get localization text for.
	 * @return the localization text for the given property, or a placeholder if the localization
	 * is null.
	 */
	default @NotNull String localizationText(Property property) {
		if (localization(property) == null) return "[null]";
		return localization(property).text();
	}

	default @NotNull String localizationStatus(Property property) {
		if (localization(property) == null) return "[null]";
		return localization(property).status().toString();
	}

	// todo may bring back at some point
//	/**
//	 * Sets the localization for the given property to the new value, or creates a new localization if none exists.
//	 * @param property the localizable property to set.
//	 * @param text the new localization text.
//\	 */
//	default void setLocalization(Property property, String text) {
//		setLocalization(property, text, primaryLocalizationFile());
//	}

//	default File primaryLocalizationFile() {
//		var localizableGroup = getLocalizableGroup();
//		return localizableGroup.stream()
//				.flatMap(localizable -> localizable.getLocalizationKeys().stream())
//				.map(LocalizationManager::getLocalizationFile)
//				.filter(Objects::nonNull)
//				.findFirst().orElse(null);
//	}
default File primaryLocalizationFile() {
	var localizableGroup = getLocalizableGroup();
	for (Object localizable : localizableGroup) {
		if (!(localizable instanceof Localizable)) {
			throw new ClassCastException("Element is not an instance of Localizable: " + localizable.getClass());
		}
		System.out.println("Localizable element: " + localizable.getClass());
	}
	return localizableGroup.stream()
			.flatMap(localizable -> localizable.getLocalizationKeys().stream())
			.map(LocalizationManager::getLocalizationFile)
			.filter(Objects::nonNull)
			.findFirst().orElse(null);
}

	/**
	 * Gets the localizable group of objects that this object is a part of.
	 * @return the localizable group of objects.
	 */
	@NotNull Collection<? extends Localizable> getLocalizableGroup();

	/**
	 * Asks the user to determine the localization file to use.
	 * @return the localization file to use.
	 */
	default File askUserForLocalizationFile() {
		File initialDirectory = HOIIVFile.mod_localization_folder;
		File file = FXWindow.openChooser(initialDirectory, false);
		if (file == null) file = new File(HOIIVFile.mod_localization_folder, "HOIUtils_extra_localization.yml");
		return file;
	}

	/**
	 * Sets the localization for the given property to the new value, or creates a new localization if none exists.
	 * @param property the localizable property to set.
	 * @param text the new localization text.
	 * @param file the file the localization belongs in.
	 */
	default void setLocalization(Property property, String text, @NotNull File file) {
		setLocalization(property, null, text, file);
	}

	/**
	 * Sets the localization for the given property to the new value, or creates a new localization if none exists.
	 * @param property the localizable property to set.
	 * @param version the localization version number
	 * @param text the new localization text.
	 * @param file the file the localization belongs in.
	 */
	default void setLocalization(Property property, Integer version, String text, @NotNull File file) {
		String key = getLocalizableProperties().get(property);
		LocalizationManager.get().setLocalization(key, version, text, file);
	}

	/**
	 * Sets the localization for the given property to the new value.
	 * @param property the localizable property to set.
	 * @param text the new localization text.
	 */
	default void replaceLocalization(Property property, String text) {
		String key = getLocalizableProperties().get(property);
		LocalizationManager.get().replaceLocalization(key, text);
	}

}

File: src/main/scala/com/hoi4utils/clausewitz/localization/Localizable.scala
package com.hoi4utils.clausewitz.localization

import com.hoi4utils.clausewitz.HOIIVFiles
import com.hoi4utils.clausewitz.localization.*
import com.hoi4utils.ui.JavaFXUIManager
import org.jetbrains.annotations.NotNull

import java.io.File
import scala.collection.mutable
import scala.collection.mutable.{HashMap, Map}

/**
 * Interface for Clausewitz-localizable objects. This interface is specifically related to
 * Clausewitz-engine localization, and not general localization/i18n
 */
trait Localizable {
  /**
   * Default method to get the number of localizable properties.
   *
   * @note this method should not usually be overwritten.
   * @return the number of localizable properties.
   */
  def numLocalizableProperties: Int = getLocalizableProperties.size

  /**
   * Default method to get the localizable property identifiers and keys.
   *
   * @return a map of localizable property identifiers and keys.
   */
  def getLocalizableProperties: mutable.Map[Property, String]
  
  def localizableProperty(property: Property): Option[String] = getLocalizableProperties.get(property)

  def getLocalizationKeys: Iterable[String] = getLocalizableProperties.values

  /**
   * Add a localizable property using the specified key.
   *
   * @param property        the localized property to add.
   * @param localizationKey the property localization key to add.
   */
  def addLocalizableProperty(property: Property, localizationKey: String): Unit = {
    getLocalizableProperties.put(property, localizationKey)
  }

  /**
   * Default method to clear the localizable properties.
   */
  def clearLocalizableProperties(): Unit = {
    getLocalizableProperties.clear()
  }

  def getLocalization: Iterable[Localization] = LocalizationManager.getAll(getLocalizationKeys)

  /**
   * Gets the localization for the given property.
   *
   * @param property the localizable property to get localization for.
   * @return the localization for the given property.
   */
  def localization(property: Property): Option[Localization] = {
    val key = getLocalizableProperties.get(property)
    key match {
      case Some(k) => LocalizationManager.get(k)
      case None => None
    }
  }

  /**
   * Gets the localization text for the given property.
   *
   * @param property the localizable property to get localization text for.
   * @return the localization text for the given property, or a placeholder if the localization
   *         is null.
   */
  def localizationText(property: Property): String = {
    localization(property) match {
      case Some(l) => l.text
      case None => "[null]"
    }
  }

  def localizationStatus(property: Property): Localization.Status = {
    localization(property) match {
      case Some(l) => l.status
      case None => Localization.Status.MISSING
    }
  }

  // todo may bring back at some point
  //	/**
  //	 * Sets the localization for the given property to the new value, or creates a new localization if none exists.
  //	 * @param property the localizable property to set.
  //	 * @param text the new localization text.
  //\	 */
  //	default void setLocalization(Property property, String text) {
  //		setLocalization(property, text, primaryLocalizationFile());
  //	}

  def primaryLocalizationFile: Option[File] = {
    val localizableGroup = getLocalizableGroup
    localizableGroup.flatMap(ll => ll.getLocalizationKeys).map(LocalizationManager.getLocalizationFile).filter(_ != null).headOption match {
      case Some(f) => Some(f)
      case None => None
    }
  }

  /**
   * Gets the localizable group of objects that this object is a part of.
   *
   * @return the localizable group of objects.
   */
  def getLocalizableGroup: Iterable[_ <: Localizable]

  /**
   * Asks the user to determine the localization file to use.
   *
   * @return the localization file to use.
   */
  def askUserForLocalizationFile: File = {
    val initialDirectory = HOIIVFiles.Mod.localization_folder
    var file = JavaFXUIManager.openChooser(initialDirectory, false)
    if (file == null) file = new File(HOIIVFiles.Mod.localization_folder, "HOIUtils_extra_localization.yml")
    file
  }

  /**
   * Sets the localization for the given property to the new value, or creates a new localization if none exists.
   *
   * @param property the localizable property to set.
   * @param text     the new localization text.
   * @param file     the file the localization belongs in.
   */
  def setLocalization(property: Property, text: String, file: File): Unit = {
    setLocalization(property, null, text, file)
  }

  /**
   * Sets the localization for the given property to the new value, or creates a new localization if none exists.
   *
   * @param property the localizable property to set.
   * @param version  the localization version number
   * @param text     the new localization text.
   * @param file     the file the localization belongs in.
   */
  def setLocalization(property: Property, version: Integer, text: String, file: File): Unit = {
    val key = getLocalizableProperties.get(property)
    key match {
      case Some(k) =>
        LocalizationManager.get.setLocalization(k, version, text, file)
      case None => //LocalizationManager.get.addLocalization(text, file)
    }
  }

  /**
   * Sets the localization for the given property to the new value.
   *
   * @param property the localizable property to set.
   * @param text     the new localization text.
   */
  def replaceLocalization(property: Property, text: String): Unit = {
    val key = getLocalizableProperties.get(property)
    key match {
      case Some(k) => LocalizationManager.get.replaceLocalization(k, text)
      case None =>
    }
  }
  
}
